src/berryimu.py

```diff
@@ -0,0 +1,243 @@
+    ) -> Tuple[GeoPoint, float]:
+         raise NotImplementedError
+
+    async def get_linear_velocity(self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs) -> Vector3:
+          raise NotImplementedError
+    
+    async def get_angular_velocity(self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs) -> Vector3:
+          return self.velocity
+
+    async def get_linear_acceleration(
+        self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs
+    ) -> Vector3:
+          return self.acceleration
+
+    async def get_compass_heading(self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs) -> float:
+         return self.compass_heading
+
+    async def get_orientation(self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs) -> Orientation:
+        return self.orientation
+
+    async def get_properties(self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs) -> MovementSensor.Properties:
+        return MovementSensor.Properties(linear_velocity_supported=False,
+                                         angular_velocity_supported=True,
+                                         orientation_supported=True,
+                                         position_supported=False,
+                                         compass_heading_supported=True,
+                                         linear_acceleration_supported=True)
+       
+    async def get_accuracy(
+        self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs
+    ) -> Mapping[str, float]:
+        raise NotImplementedError
+    
+    def read(self):
+        # This loop must run for the same amount of time each loop to accurately complete the complementary filter. 
+        while True:
+            now = time.time() 
+            self.velocity = self.read_angular_velocity()
+            self.compass_heading = self.calculate_compass_heading()
+            self.acceleration = self.get_acceleration()
+            self.orientation = self.calculate_orientation()
+            elapsed = time.time() - now
+            time.sleep(dt - elapsed) # full iteration takes 80 ms
+
+    def calculate_compass_heading(self):
+        raw_data = self.i2cbus.read_i2c_block_data(constants.MAG_ADDRESS,  constants.OUT_X_L, 6)
+        values = utils.parse_output(raw_data)
+        if self.calibrate:
+            # hard iron distortion correction
+            values.x -= (self.hard_iron_x_min + self.hard_iron_x_max) /2
+            values.y -= (self.hard_iron_y_min + self.hard_iron_y_max) /2
+
+            # #soft iron distortion correction 
+            scale_y = self.soft_iron_y_max - self.soft_iron_y_min
+            scale_x = self.soft_iron_x_max - self.soft_iron_x_min 
+            ratio = scale_y / scale_x
```

Comment:
will scale_x ever be 0? if so we should guard against that

Reply:
only if the max and min are equal, which would only happen if something went very wrong during the calibration.

Reply:
then let's just make sure to validate against that case in the validator function and tell the user to recalibrate in that instance 👯